<!DOCTYPE html>
<html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8"/>

<title>clonotypeR: Identify and analyse B and T cell receptors at a high throughput</title>

<!-- Styles for R syntax highlighter -->
<style type="text/css">
   pre .operator,
   pre .paren {
     color: rgb(104, 118, 135)
   }

   pre .literal {
     color: rgb(88, 72, 246)
   }

   pre .number {
     color: rgb(0, 0, 205);
   }

   pre .comment {
     color: rgb(76, 136, 107);
   }

   pre .keyword {
     color: rgb(0, 0, 255);
   }

   pre .identifier {
     color: rgb(0, 0, 0);
   }

   pre .string {
     color: rgb(3, 106, 7);
   }
</style>

<!-- R syntax highlighter -->
<script type="text/javascript">
var hljs=new function(){function m(p){return p.replace(/&/gm,"&amp;").replace(/</gm,"&lt;")}function f(r,q,p){return RegExp(q,"m"+(r.cI?"i":"")+(p?"g":""))}function b(r){for(var p=0;p<r.childNodes.length;p++){var q=r.childNodes[p];if(q.nodeName=="CODE"){return q}if(!(q.nodeType==3&&q.nodeValue.match(/\s+/))){break}}}function h(t,s){var p="";for(var r=0;r<t.childNodes.length;r++){if(t.childNodes[r].nodeType==3){var q=t.childNodes[r].nodeValue;if(s){q=q.replace(/\n/g,"")}p+=q}else{if(t.childNodes[r].nodeName=="BR"){p+="\n"}else{p+=h(t.childNodes[r])}}}if(/MSIE [678]/.test(navigator.userAgent)){p=p.replace(/\r/g,"\n")}return p}function a(s){var r=s.className.split(/\s+/);r=r.concat(s.parentNode.className.split(/\s+/));for(var q=0;q<r.length;q++){var p=r[q].replace(/^language-/,"");if(e[p]){return p}}}function c(q){var p=[];(function(s,t){for(var r=0;r<s.childNodes.length;r++){if(s.childNodes[r].nodeType==3){t+=s.childNodes[r].nodeValue.length}else{if(s.childNodes[r].nodeName=="BR"){t+=1}else{if(s.childNodes[r].nodeType==1){p.push({event:"start",offset:t,node:s.childNodes[r]});t=arguments.callee(s.childNodes[r],t);p.push({event:"stop",offset:t,node:s.childNodes[r]})}}}}return t})(q,0);return p}function k(y,w,x){var q=0;var z="";var s=[];function u(){if(y.length&&w.length){if(y[0].offset!=w[0].offset){return(y[0].offset<w[0].offset)?y:w}else{return w[0].event=="start"?y:w}}else{return y.length?y:w}}function t(D){var A="<"+D.nodeName.toLowerCase();for(var B=0;B<D.attributes.length;B++){var C=D.attributes[B];A+=" "+C.nodeName.toLowerCase();if(C.value!==undefined&&C.value!==false&&C.value!==null){A+='="'+m(C.value)+'"'}}return A+">"}while(y.length||w.length){var v=u().splice(0,1)[0];z+=m(x.substr(q,v.offset-q));q=v.offset;if(v.event=="start"){z+=t(v.node);s.push(v.node)}else{if(v.event=="stop"){var p,r=s.length;do{r--;p=s[r];z+=("</"+p.nodeName.toLowerCase()+">")}while(p!=v.node);s.splice(r,1);while(r<s.length){z+=t(s[r]);r++}}}}return z+m(x.substr(q))}function j(){function q(x,y,v){if(x.compiled){return}var u;var s=[];if(x.k){x.lR=f(y,x.l||hljs.IR,true);for(var w in x.k){if(!x.k.hasOwnProperty(w)){continue}if(x.k[w] instanceof Object){u=x.k[w]}else{u=x.k;w="keyword"}for(var r in u){if(!u.hasOwnProperty(r)){continue}x.k[r]=[w,u[r]];s.push(r)}}}if(!v){if(x.bWK){x.b="\\b("+s.join("|")+")\\s"}x.bR=f(y,x.b?x.b:"\\B|\\b");if(!x.e&&!x.eW){x.e="\\B|\\b"}if(x.e){x.eR=f(y,x.e)}}if(x.i){x.iR=f(y,x.i)}if(x.r===undefined){x.r=1}if(!x.c){x.c=[]}x.compiled=true;for(var t=0;t<x.c.length;t++){if(x.c[t]=="self"){x.c[t]=x}q(x.c[t],y,false)}if(x.starts){q(x.starts,y,false)}}for(var p in e){if(!e.hasOwnProperty(p)){continue}q(e[p].dM,e[p],true)}}function d(B,C){if(!j.called){j();j.called=true}function q(r,M){for(var L=0;L<M.c.length;L++){if((M.c[L].bR.exec(r)||[null])[0]==r){return M.c[L]}}}function v(L,r){if(D[L].e&&D[L].eR.test(r)){return 1}if(D[L].eW){var M=v(L-1,r);return M?M+1:0}return 0}function w(r,L){return L.i&&L.iR.test(r)}function K(N,O){var M=[];for(var L=0;L<N.c.length;L++){M.push(N.c[L].b)}var r=D.length-1;do{if(D[r].e){M.push(D[r].e)}r--}while(D[r+1].eW);if(N.i){M.push(N.i)}return f(O,M.join("|"),true)}function p(M,L){var N=D[D.length-1];if(!N.t){N.t=K(N,E)}N.t.lastIndex=L;var r=N.t.exec(M);return r?[M.substr(L,r.index-L),r[0],false]:[M.substr(L),"",true]}function z(N,r){var L=E.cI?r[0].toLowerCase():r[0];var M=N.k[L];if(M&&M instanceof Array){return M}return false}function F(L,P){L=m(L);if(!P.k){return L}var r="";var O=0;P.lR.lastIndex=0;var M=P.lR.exec(L);while(M){r+=L.substr(O,M.index-O);var N=z(P,M);if(N){x+=N[1];r+='<span class="'+N[0]+'">'+M[0]+"</span>"}else{r+=M[0]}O=P.lR.lastIndex;M=P.lR.exec(L)}return r+L.substr(O,L.length-O)}function J(L,M){if(M.sL&&e[M.sL]){var r=d(M.sL,L);x+=r.keyword_count;return r.value}else{return F(L,M)}}function I(M,r){var L=M.cN?'<span class="'+M.cN+'">':"";if(M.rB){y+=L;M.buffer=""}else{if(M.eB){y+=m(r)+L;M.buffer=""}else{y+=L;M.buffer=r}}D.push(M);A+=M.r}function G(N,M,Q){var R=D[D.length-1];if(Q){y+=J(R.buffer+N,R);return false}var P=q(M,R);if(P){y+=J(R.buffer+N,R);I(P,M);return P.rB}var L=v(D.length-1,M);if(L){var O=R.cN?"</span>":"";if(R.rE){y+=J(R.buffer+N,R)+O}else{if(R.eE){y+=J(R.buffer+N,R)+O+m(M)}else{y+=J(R.buffer+N+M,R)+O}}while(L>1){O=D[D.length-2].cN?"</span>":"";y+=O;L--;D.length--}var r=D[D.length-1];D.length--;D[D.length-1].buffer="";if(r.starts){I(r.starts,"")}return R.rE}if(w(M,R)){throw"Illegal"}}var E=e[B];var D=[E.dM];var A=0;var x=0;var y="";try{var s,u=0;E.dM.buffer="";do{s=p(C,u);var t=G(s[0],s[1],s[2]);u+=s[0].length;if(!t){u+=s[1].length}}while(!s[2]);if(D.length>1){throw"Illegal"}return{r:A,keyword_count:x,value:y}}catch(H){if(H=="Illegal"){return{r:0,keyword_count:0,value:m(C)}}else{throw H}}}function g(t){var p={keyword_count:0,r:0,value:m(t)};var r=p;for(var q in e){if(!e.hasOwnProperty(q)){continue}var s=d(q,t);s.language=q;if(s.keyword_count+s.r>r.keyword_count+r.r){r=s}if(s.keyword_count+s.r>p.keyword_count+p.r){r=p;p=s}}if(r.language){p.second_best=r}return p}function i(r,q,p){if(q){r=r.replace(/^((<[^>]+>|\t)+)/gm,function(t,w,v,u){return w.replace(/\t/g,q)})}if(p){r=r.replace(/\n/g,"<br>")}return r}function n(t,w,r){var x=h(t,r);var v=a(t);var y,s;if(v){y=d(v,x)}else{return}var q=c(t);if(q.length){s=document.createElement("pre");s.innerHTML=y.value;y.value=k(q,c(s),x)}y.value=i(y.value,w,r);var u=t.className;if(!u.match("(\\s|^)(language-)?"+v+"(\\s|$)")){u=u?(u+" "+v):v}if(/MSIE [678]/.test(navigator.userAgent)&&t.tagName=="CODE"&&t.parentNode.tagName=="PRE"){s=t.parentNode;var p=document.createElement("div");p.innerHTML="<pre><code>"+y.value+"</code></pre>";t=p.firstChild.firstChild;p.firstChild.cN=s.cN;s.parentNode.replaceChild(p.firstChild,s)}else{t.innerHTML=y.value}t.className=u;t.result={language:v,kw:y.keyword_count,re:y.r};if(y.second_best){t.second_best={language:y.second_best.language,kw:y.second_best.keyword_count,re:y.second_best.r}}}function o(){if(o.called){return}o.called=true;var r=document.getElementsByTagName("pre");for(var p=0;p<r.length;p++){var q=b(r[p]);if(q){n(q,hljs.tabReplace)}}}function l(){if(window.addEventListener){window.addEventListener("DOMContentLoaded",o,false);window.addEventListener("load",o,false)}else{if(window.attachEvent){window.attachEvent("onload",o)}else{window.onload=o}}}var e={};this.LANGUAGES=e;this.highlight=d;this.highlightAuto=g;this.fixMarkup=i;this.highlightBlock=n;this.initHighlighting=o;this.initHighlightingOnLoad=l;this.IR="[a-zA-Z][a-zA-Z0-9_]*";this.UIR="[a-zA-Z_][a-zA-Z0-9_]*";this.NR="\\b\\d+(\\.\\d+)?";this.CNR="\\b(0[xX][a-fA-F0-9]+|(\\d+(\\.\\d*)?|\\.\\d+)([eE][-+]?\\d+)?)";this.BNR="\\b(0b[01]+)";this.RSR="!|!=|!==|%|%=|&|&&|&=|\\*|\\*=|\\+|\\+=|,|\\.|-|-=|/|/=|:|;|<|<<|<<=|<=|=|==|===|>|>=|>>|>>=|>>>|>>>=|\\?|\\[|\\{|\\(|\\^|\\^=|\\||\\|=|\\|\\||~";this.ER="(?![\\s\\S])";this.BE={b:"\\\\.",r:0};this.ASM={cN:"string",b:"'",e:"'",i:"\\n",c:[this.BE],r:0};this.QSM={cN:"string",b:'"',e:'"',i:"\\n",c:[this.BE],r:0};this.CLCM={cN:"comment",b:"//",e:"$"};this.CBLCLM={cN:"comment",b:"/\\*",e:"\\*/"};this.HCM={cN:"comment",b:"#",e:"$"};this.NM={cN:"number",b:this.NR,r:0};this.CNM={cN:"number",b:this.CNR,r:0};this.BNM={cN:"number",b:this.BNR,r:0};this.inherit=function(r,s){var p={};for(var q in r){p[q]=r[q]}if(s){for(var q in s){p[q]=s[q]}}return p}}();hljs.LANGUAGES.cpp=function(){var a={keyword:{"false":1,"int":1,"float":1,"while":1,"private":1,"char":1,"catch":1,"export":1,virtual:1,operator:2,sizeof:2,dynamic_cast:2,typedef:2,const_cast:2,"const":1,struct:1,"for":1,static_cast:2,union:1,namespace:1,unsigned:1,"long":1,"throw":1,"volatile":2,"static":1,"protected":1,bool:1,template:1,mutable:1,"if":1,"public":1,friend:2,"do":1,"return":1,"goto":1,auto:1,"void":2,"enum":1,"else":1,"break":1,"new":1,extern:1,using:1,"true":1,"class":1,asm:1,"case":1,typeid:1,"short":1,reinterpret_cast:2,"default":1,"double":1,register:1,explicit:1,signed:1,typename:1,"try":1,"this":1,"switch":1,"continue":1,wchar_t:1,inline:1,"delete":1,alignof:1,char16_t:1,char32_t:1,constexpr:1,decltype:1,noexcept:1,nullptr:1,static_assert:1,thread_local:1,restrict:1,_Bool:1,complex:1},built_in:{std:1,string:1,cin:1,cout:1,cerr:1,clog:1,stringstream:1,istringstream:1,ostringstream:1,auto_ptr:1,deque:1,list:1,queue:1,stack:1,vector:1,map:1,set:1,bitset:1,multiset:1,multimap:1,unordered_set:1,unordered_map:1,unordered_multiset:1,unordered_multimap:1,array:1,shared_ptr:1}};return{dM:{k:a,i:"</",c:[hljs.CLCM,hljs.CBLCLM,hljs.QSM,{cN:"string",b:"'\\\\?.",e:"'",i:"."},{cN:"number",b:"\\b(\\d+(\\.\\d*)?|\\.\\d+)(u|U|l|L|ul|UL|f|F)"},hljs.CNM,{cN:"preprocessor",b:"#",e:"$"},{cN:"stl_container",b:"\\b(deque|list|queue|stack|vector|map|set|bitset|multiset|multimap|unordered_map|unordered_set|unordered_multiset|unordered_multimap|array)\\s*<",e:">",k:a,r:10,c:["self"]}]}}}();hljs.LANGUAGES.r={dM:{c:[hljs.HCM,{cN:"number",b:"\\b0[xX][0-9a-fA-F]+[Li]?\\b",e:hljs.IMMEDIATE_RE,r:0},{cN:"number",b:"\\b\\d+(?:[eE][+\\-]?\\d*)?L\\b",e:hljs.IMMEDIATE_RE,r:0},{cN:"number",b:"\\b\\d+\\.(?!\\d)(?:i\\b)?",e:hljs.IMMEDIATE_RE,r:1},{cN:"number",b:"\\b\\d+(?:\\.\\d*)?(?:[eE][+\\-]?\\d*)?i?\\b",e:hljs.IMMEDIATE_RE,r:0},{cN:"number",b:"\\.\\d+(?:[eE][+\\-]?\\d*)?i?\\b",e:hljs.IMMEDIATE_RE,r:1},{cN:"keyword",b:"(?:tryCatch|library|setGeneric|setGroupGeneric)\\b",e:hljs.IMMEDIATE_RE,r:10},{cN:"keyword",b:"\\.\\.\\.",e:hljs.IMMEDIATE_RE,r:10},{cN:"keyword",b:"\\.\\.\\d+(?![\\w.])",e:hljs.IMMEDIATE_RE,r:10},{cN:"keyword",b:"\\b(?:function)",e:hljs.IMMEDIATE_RE,r:2},{cN:"keyword",b:"(?:if|in|break|next|repeat|else|for|return|switch|while|try|stop|warning|require|attach|detach|source|setMethod|setClass)\\b",e:hljs.IMMEDIATE_RE,r:1},{cN:"literal",b:"(?:NA|NA_integer_|NA_real_|NA_character_|NA_complex_)\\b",e:hljs.IMMEDIATE_RE,r:10},{cN:"literal",b:"(?:NULL|TRUE|FALSE|T|F|Inf|NaN)\\b",e:hljs.IMMEDIATE_RE,r:1},{cN:"identifier",b:"[a-zA-Z.][a-zA-Z0-9._]*\\b",e:hljs.IMMEDIATE_RE,r:0},{cN:"operator",b:"<\\-(?!\\s*\\d)",e:hljs.IMMEDIATE_RE,r:2},{cN:"operator",b:"\\->|<\\-",e:hljs.IMMEDIATE_RE,r:1},{cN:"operator",b:"%%|~",e:hljs.IMMEDIATE_RE},{cN:"operator",b:">=|<=|==|!=|\\|\\||&&|=|\\+|\\-|\\*|/|\\^|>|<|!|&|\\||\\$|:",e:hljs.IMMEDIATE_RE,r:0},{cN:"operator",b:"%",e:"%",i:"\\n",r:1},{cN:"identifier",b:"`",e:"`",r:0},{cN:"string",b:'"',e:'"',c:[hljs.BE],r:0},{cN:"string",b:"'",e:"'",c:[hljs.BE],r:0},{cN:"paren",b:"[[({\\])}]",e:hljs.IMMEDIATE_RE,r:0}]}};
hljs.initHighlightingOnLoad();
</script>

<!-- MathJax scripts -->
<script type="text/javascript" src="https://c328740.ssl.cf1.rackcdn.com/mathjax/2.0-latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML">
</script>


<style type="text/css">
body, td {
   font-family: sans-serif;
   background-color: white;
   font-size: 12px;
   margin: 8px;
}

tt, code, pre {
   font-family: 'DejaVu Sans Mono', 'Droid Sans Mono', 'Lucida Console', Consolas, Monaco, monospace;
}

h1 { 
   font-size:2.2em; 
}

h2 { 
   font-size:1.8em; 
}

h3 { 
   font-size:1.4em; 
}

h4 { 
   font-size:1.0em; 
}

h5 { 
   font-size:0.9em; 
}

h6 { 
   font-size:0.8em; 
}

a:visited {
   color: rgb(50%, 0%, 50%);
}

pre {	
   margin-top: 0;
   max-width: 95%;
   border: 1px solid #ccc;
   white-space: pre-wrap;
}

pre code {
   display: block; padding: 0.5em;
}

code.r, code.cpp {
   background-color: #F8F8F8;
}

table, td, th {
  border: none;
}

blockquote {
   color:#666666;
   margin:0;
   padding-left: 1em;
   border-left: 0.5em #EEE solid;
}

hr {
   height: 0px;
   border-bottom: none;
   border-top-width: thin;
   border-top-style: dotted;
   border-top-color: #999999;
}

@media print {
   * { 
      background: transparent !important; 
      color: black !important; 
      filter:none !important; 
      -ms-filter: none !important; 
   }

   body { 
      font-size:12pt; 
      max-width:100%; 
   }
       
   a, a:visited { 
      text-decoration: underline; 
   }

   hr { 
      visibility: hidden;
      page-break-before: always;
   }

   pre, blockquote { 
      padding-right: 1em; 
      page-break-inside: avoid; 
   }

   tr, img { 
      page-break-inside: avoid; 
   }

   img { 
      max-width: 100% !important; 
   }

   @page :left { 
      margin: 15mm 20mm 15mm 10mm; 
   }
     
   @page :right { 
      margin: 15mm 10mm 15mm 20mm; 
   }

   p, h2, h3 { 
      orphans: 3; widows: 3; 
   }

   h2, h3 { 
      page-break-after: avoid; 
   }
}
</style>



</head>

<body>
<!--
%\VignetteEngine{knitr::knitr}
%\VignetteIndexEntry{clonotypeR User's Guide}
\usepackage[utf8]{inputenc}
-->

<h1>clonotypeR: Identify and analyse B and T cell receptors at a high throughput</h1>

<pre><code>Author: Charles Plessy &lt;plessy@riken.jp&gt;
Date:   1 May 2013
</code></pre>

<p><em>clonotypeR</em> is a R package and accompanying scripts to identify and
analyse clonotypes from high-throughput T cell receptors sequence
libraries. <em>clonotypeR</em> is suited to process and organise very large
number of clonotypes, in the order of millions, typically produced by
Roche 454 instruments, and to prepare these sequences for differential
expression analysis with the typical transcriptomics tools as well as
for statistical analysis using existing R packages.</p>

<p>The home page of <em>clonotypeR</em> is <a href="http://clonotyper.branchable.com/">http://clonotyper.branchable.com/</a>.</p>

<h2>clonotypeR&#39;s workflow</h2>

<p>Typically, the user receives the output of a next-generation sequencer and runs 
some shell commands that are not part of the <em>clonotypeR</em> R package, but that are 
distributed with it on <a href="http://clonotyper.branchable.com/">http://clonotyper.branchable.com/</a>.</p>

<p>Currently, <em>clonotypeR</em> provides only support for mouse sequences, by providing
pre-formatted sequences from <code>V</code> and <code>J</code> segments.  Support of human sequences
is in preparation.  See <a href="http://clonotyper.branchable.com/references/README">http://clonotyper.branchable.com/references/README</a>
for more information.</p>

<p>The workflow presented here summarises the different commands to run.  Other
examples are available on line at
<a href="http://clonotyper.branchable.com/doc/workflow/">http://clonotyper.branchable.com/doc/workflow/</a>.</p>

<p><em>This example analysis assumes a
<a href="https://en.wikipedia.org/wiki/Operating%5Fsystem%23UNIX%5Fand%5FUNIX%2Dlike%5Foperating%5Fsystems">unix</a>
system (Linux, Mac OS, …)</em></p>

<h2>Test data</h2>

<p>ClonotypeR come with test data in the <code>R</code> package.  While it is not deep enough
to discuss about biology, you can use it to familiarise yourself with the
commands or test them.</p>

<p>The R package is loaded as usual.</p>

<pre><code class="r">library(clonotypeR)
</code></pre>

<pre><code>## Loading required package: methods
</code></pre>

<p>The data is a table of 120 clonotypes in the <code>extdata</code> folder of the package.
The command <code>read_clonotypes</code> will parse it in a data frame.  The clonotypes
are arbitrarly assigned to three libraries called <code>A</code>, <code>B</code>, and <code>C</code>.  The
<code>read_clonotypes</code> comments determines at load time if the peptidic sequence has
a stop codon or is frame-shifted, and records the information in the
<code>unproductive</code> column.</p>

<pre><code class="r">clonotypes &lt;- read_clonotypes(system.file(&quot;extdata&quot;, &quot;clonotypes.txt.gz&quot;, package = &quot;clonotypeR&quot;))
summary(clonotypes)
</code></pre>

<pre><code>##  lib                V            J          score          mapq      
##  A:40   TRAV14-2     : 1   TRAJ31 :35   Min.   : 45   Min.   :  0.0  
##  B:40   TRAV14-3     :23   TRAJ49 :10   1st Qu.:215   1st Qu.: 31.0  
##  C:40   TRAV14D-3/DV8: 1   TRAJ57 :10   Median :226   Median : 37.5  
##         TRAV14N-1    :83   TRAJ34 : 9   Mean   :222   Mean   : 36.6  
##         TRAV14N-2    :11   TRAJ27 : 8   3rd Qu.:233   3rd Qu.: 40.0  
##         TRAV14N-3    : 1   TRAJ23 : 7   Max.   :247   Max.   :218.0  
##                            (Other):41                                
##      read               dna                qual          
##  Length:120         Length:120         Length:120        
##  Class :character   Class :character   Class :character  
##  Mode  :character   Mode  :character   Mode  :character  
##                                                          
##                                                          
##                                                          
##                                                          
##      pep            unproductive   
##  Length:120         Mode :logical  
##  Class :character   FALSE:102      
##  Mode  :character   TRUE :18       
##                     NA&#39;s :0        
##                                    
##                                    
## 
</code></pre>

<p>The <code>clonotype_table</code> command counts how many times a given clonotype is found
in each library.  It can also count simpler features, in particular <code>V</code> and <code>J</code>
segments, or any combination of them.</p>

<pre><code class="r">head(clonotype_table(levels(clonotypes$lib), data = clonotypes))
</code></pre>

<pre><code>##                                A B C
## TRAV14-2 AASAHYGSSGNKLI TRAJ32 1 0 0
## TRAV14-3 AAADYSNNRLT TRAJ7     1 0 0
## TRAV14-3 AAISNYNQGKLI TRAJ23   1 0 0
## TRAV14-3 AAMNQGGSAKLI TRAJ57   1 1 0
## TRAV14-3 AASAVDYANKMI TRAJ47   1 0 0
## TRAV14-3 AATVDSNYQLI TRAJ33    1 0 0
</code></pre>

<pre><code class="r">head(clonotype_table(levels(clonotypes$lib), &quot;V&quot;, data = clonotypes))
</code></pre>

<pre><code>##                A  B  C
## TRAV14-2       1  0  0
## TRAV14-3       6  7  5
## TRAV14N-1     23 24 24
## TRAV14N-2      2  3  4
## TRAV14N-3      1  0  0
## TRAV14D-3/DV8  0  0  1
</code></pre>

<pre><code class="r">head(clonotype_table(levels(clonotypes$lib), &quot;J&quot;, data = clonotypes))
</code></pre>

<pre><code>##        A  B  C
## TRAJ13 1  1  0
## TRAJ15 1  0  1
## TRAJ23 5  0  1
## TRAJ26 2  2  0
## TRAJ27 2  2  3
## TRAJ31 9 12 12
</code></pre>

<p>ClonotypeR provides other functions for further analysis.  <code>yassai_identifier</code>
calculates a unique indentifier using the <code>V</code>, <code>J</code>, peptidic and nucleotidic
information, following the work of <a href="http://dx.doi.org/10.1007/s00251-009-0383-x">Yassai et al</a>.</p>

<pre><code class="r"># Unique identifier
head(yassai_identifier(clonotypes))
</code></pre>

<pre><code>## [1] &quot;aAYt.22A14N2A49L10&quot;     &quot;aYt.2A14N2A49L10&quot;      
## [3] &quot;aGRt.34A14N1A40L11&quot;     &quot;aADTIt.1221A14N1A27L11&quot;
## [5] &quot;AEMn.411A14N1A23L11&quot;    &quot;sAHy.11A14-2A32L14&quot;
</code></pre>

<p><code>unique_clonotypes</code> and <code>common_clonotypes</code> are typically used when comparing libraries.</p>

<pre><code class="r">clonotypes &lt;- clonotype_table(levels(clonotypes$lib), data = clonotypes)

# First six clonotypes of library C
head(unique_clonotypes(&quot;C&quot;, data = clonotypes))
</code></pre>

<pre><code>## [1] &quot;TRAV14N-1 AASDTNTGKLT TRAJ27&quot; &quot;TRAV14N-1 AASNSNNRIF TRAJ31&quot; 
## [3] &quot;TRAV14N-2 AAYTGYQNFY TRAJ49&quot;  &quot;TRAV14-3 AASYGSSGNKLI TRAJ32&quot;
## [5] &quot;TRAV14N-1 AALNSNNRIF TRAJ31&quot;  &quot;TRAV14N-1 AAPSNTNKVV TRAJ34&quot;
</code></pre>

<pre><code class="r"># Count clonotypes found in library A, and B or C.
length(common_clonotypes(group1 = &quot;A&quot;, group2 = c(&quot;B&quot;, &quot;C&quot;), data = clonotypes))
</code></pre>

<pre><code>## [1] 5
</code></pre>

<pre><code class="r"># Matrix of numbers of common clonotypes
common_clonotypes(data = clonotypes)
</code></pre>

<pre><code>##    A  B  C
## A 26  2  3
## B  2 27  4
## C  3  4 27
</code></pre>

<p>With deeper data, a typical follow-up would be to identify differentially
represnted clonotypes between libraries, for instance with the <code>edgeR</code> package,
or to calculate distance between libraries, for instance with the
<code>vegan</code> package.</p>

<h2>Example data (3 × 2,000 reads)</h2>

<p>The data provided on-line at <a href="http://clonotyper.branchable.com/example_data/">http://clonotyper.branchable.com/example_data/</a>
is a sub-sample of three sequence librairies of mouse T cell receptors α (2,000
reads each) made on the 454 Titanium or the 454 junior platforms.  The original
libraries will be deposited in public databanks after publication in a
peer-reviewed journal.</p>

<p>These example libraries are called <code>A</code>, <code>B</code> and <code>C</code>, and are in FASTQ format,
with entries like the following (the sequence was truncated for the convenience
of the display).</p>

<pre><code>@HKTLYLP01B0MTM
gactGTCCATCTTCCTTTTATCGGACACTGAAGTATGGATATCAGAAGTGCAgggccttcccacgggaacg
+
IIIIIIIIIIIHHFF::::G&amp;gt;IIIGGGIIIIIIIIIGGIIIIIIFEBDCDC&amp;lt;//-5522------
</code></pre>

<h3>Detection of V segments</h3>

<p>Run the command <code>clonotyper detect A.fastq</code> in the same directory as a copy of
the file <code>A.fastq</code>.</p>

<p>The result is stored in a temporary directory called <code>extraction_files</code>, that
will be created if it does not already exist.</p>

<p><code>clonotyper detect</code> compares the sequences to the reference V segments using
BWA, and produces output like the following.</p>

<pre><code>[bsw2_aln] read 2000 sequences/pairs (843395 bp)...
[samopen] SAM header is present: 167 sequences.
[main] Version: 0.6.2-r126
[main] CMD: bwa bwasw -t8 /usr/share/clonotypeR/references/V/index A.fastq
[main] Real time: 1.099 sec; CPU: 8.225 sec
</code></pre>

<p>This indicates that 2,000 reads have been processed, representing 843,395 base
pairs in total.  There were 167 reference V segments, and the version number of
BWA was <code>0.6.2-r126</code>.  The whole process took less than 10 seconds.</p>

<p>Process the example libraries <code>B</code> and <code>C</code> similarly with the commands
<code>clonotyper detect C.fastq</code> and <code>clonotyper detect C.fastq</code>.</p>

<h3>Extraction of CDR3 regions</h3>

<p>Run the command <code>clonotyper extract A</code> in the same directory as where you ran
<code>clonotyper detect A.fastq</code>.  The result is a table stored in a directory
called <code>clonotypes</code>, that will be created if it does not already exist.</p>

<p>The output is quite voluminous, and indicates which <em>V</em> / <em>J</em> combinations are
being found, like on the following.</p>

<pre><code>TRAV14-3    233
    TRAJ61  0
    TRAJ60  0
    TRAJ59  0
    TRAJ58  1
    TRAJ57  39
    TRAJ56  2
    TRAJ55  0
</code></pre>

<p>The format of the table is explained in the manual page of the function
<code>read_clonotypes()</code> of the R package.</p>

<p>For each library (<code>A</code>, <code>B</code> and <code>C</code>), one file is available in the <code>clonotypes</code>
directory.  With BWA <code>0.6.2-r126</code>, the following numbers of clonotypes are found.</p>

<pre><code>  1072 clonotypes/A.tsv
   924 clonotypes/B.tsv
   689 clonotypes/C.tsv
</code></pre>

<p>The files need to be concatenated before analysis in <code>R</code>, with the following command.</p>

<pre><code>find clonotypes/ -name &#39;*tsv&#39; | xargs cat &gt; clonotypes.tsv
</code></pre>

<p>A copy of the result file is provided in <code>inst/extdata/clonotypes2.tsv.xz</code> for convenience.</p>

<h3>Data analysis in R</h3>

<p>Load the clonotypeR library:</p>

<pre><code>library(clonotypeR)
</code></pre>

<p>Load the data in a R object called <em>clonotypes</em>:</p>

<pre><code>clonotypes &lt;- read_clonotypes(&#39;clonotypes.tsv&#39;)
</code></pre>

<p>Alternatively, you can load the convenience copy from
<code>inst/extdata/clonotypes2.tsv.xz</code> (see above).</p>

<pre><code class="r">clonotypes &lt;- read_clonotypes(system.file(&quot;extdata&quot;, &quot;clonotypes2.tsv.xz&quot;, package = &quot;clonotypeR&quot;))
</code></pre>

<p>The command <code>summary(clonotypes)</code> already provides useful information.</p>

<pre><code class="r">summary(clonotypes)
</code></pre>

<pre><code>##  lib                  V              J            score    
##  A:1114   TRAV14-1     :1000   TRAJ31 : 403   Min.   : 44  
##  B: 996   TRAV14-2     : 247   TRAJ23 : 281   1st Qu.:214  
##  C: 716   TRAV14-3     : 265   TRAJ22 : 279   Median :225  
##           TRAV14D-3/DV8: 255   TRAJ37 : 160   Mean   :219  
##           TRAV14N-1    : 624   TRAJ34 : 147   3rd Qu.:232  
##           TRAV14N-2    : 250   TRAJ40 : 112   Max.   :248  
##           TRAV14N-3    : 185   (Other):1444                
##       mapq           read               dna                qual          
##  Min.   :  0.0   Length:2826        Length:2826        Length:2826       
##  1st Qu.: 31.0   Class :character   Class :character   Class :character  
##  Median : 38.0   Mode  :character   Mode  :character   Mode  :character  
##  Mean   : 63.5                                                           
##  3rd Qu.: 56.8                                                           
##  Max.   :223.0                                                           
##                                                                          
##      pep            unproductive   
##  Length:2826        Mode :logical  
##  Class :character   FALSE:2223     
##  Mode  :character   TRUE :603      
##                     NA&#39;s :0        
##                                    
##                                    
## 
</code></pre>

<p>Identify unique clonotypes, count their sequences in the libraries <code>A</code>, <code>B</code> and <code>C</code>, and store the result as a table arbitrarly named <code>abc</code>. </p>

<pre><code class="r">abc &lt;- clonotype_table(c(&quot;A&quot;, &quot;B&quot;, &quot;C&quot;), data = clonotypes)
head(abc)
</code></pre>

<pre><code>##                             A B C
## TRAV14-1 AAASSGSWQLI TRAJ22 1 0 0
## TRAV14-1 AACNNRIF TRAJ31    1 0 0
## TRAV14-1 AAGAKLT TRAJ39     3 0 0
## TRAV14-1 AAGGSWQLI TRAJ22   1 0 0
## TRAV14-1 AAGTNTGKLT TRAJ27  1 0 0
## TRAV14-1 AAHDTNAYKVI TRAJ30 1 0 0
</code></pre>

<pre><code class="r">summary(abc)
</code></pre>

<pre><code>##        A                B                C        
##  Min.   : 0.000   Min.   : 0.000   Min.   :  0.0  
##  1st Qu.: 0.000   1st Qu.: 0.000   1st Qu.:  0.0  
##  Median : 0.000   Median : 0.000   Median :  0.0  
##  Mean   : 0.764   Mean   : 0.679   Mean   :  0.5  
##  3rd Qu.: 1.000   3rd Qu.: 1.000   3rd Qu.:  0.0  
##  Max.   :18.000   Max.   :22.000   Max.   :124.0
</code></pre>

<p>The summary shows that the most frequent clonotype is in <code>C</code>. Using <code>R</code> <a href="http://cran.r-project.org/doc/manuals/R-intro.html#Index-vectors">index vectors</a>, we can see that its CDR3 sequence is AASDSNNRIF and that it was not found in the other libraries.</p>

<pre><code class="r">abc[abc$C == 124, ]
</code></pre>

<pre><code>##                             A B   C
## TRAV14N-1 AASDSNNRIF TRAJ31 0 0 124
</code></pre>

<p>The <code>clonotype_table</code> function can also produce a count table for and combination of <em>V</em>, <em>CDR3</em> or <em>J</em> segments.</p>

<pre><code class="r">clonotype_table(c(&quot;A&quot;, &quot;B&quot;, &quot;C&quot;), &quot;V&quot;, data = clonotypes)
</code></pre>

<pre><code>##                 A   B   C
## TRAV14-1      245 523   0
## TRAV14-2      134  65   0
## TRAV14-3       83  11 117
## TRAV14D-3/DV8 145  51   4
## TRAV14N-1      78  26 399
## TRAV14N-2      83  63  52
## TRAV14N-3     102  34   2
</code></pre>

<pre><code class="r">head(clonotype_table(c(&quot;A&quot;, &quot;B&quot;, &quot;C&quot;), c(&quot;V&quot;, &quot;J&quot;), data = clonotypes))
</code></pre>

<pre><code>##                  A  B C
## TRAV14-1 TRAJ11  1  1 0
## TRAV14-1 TRAJ12  2  2 0
## TRAV14-1 TRAJ13  2  1 0
## TRAV14-1 TRAJ15 11 11 0
## TRAV14-1 TRAJ16  4  3 0
## TRAV14-1 TRAJ18  1 22 0
</code></pre>

</body>

</html>
