<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html><head><title>R: Reads a clonotype_table and returns a data frame.</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8">
<link rel="stylesheet" type="text/css" href="R.css">
</head><body>

<table width="100%" summary="page for read_clonotypes {clonotypeR}"><tr><td>read_clonotypes {clonotypeR}</td><td align="right">R Documentation</td></tr></table>

<h2>
Reads a clonotype_table and returns a data frame.
</h2>

<h3>Description</h3>

<p>Reads a clonotype_table in a TAB-separated or OSCT format, and returns a data frame
that has eight columns, for library name, V and J segments names, sequence read
identifier, DNA, sequence quality, aminoacid sequence of the CDR3 region, mark
for unproductive recombinations, and mark for ambiguous sequences.
</p>


<h3>Usage</h3>

<pre>
read_clonotypes(filename, scores=TRUE, ...)
</pre>


<h3>Arguments</h3>

<table summary="R argblock">
<tr valign="top"><td><code>filename</code></td>
<td>
<p>Path to the tabulation-delimited text file containing the extracted clonotypes.</p>
</td></tr>
<tr valign="top"><td><code>scores</code></td>
<td>
<p>Set to false to load legacy data that did not contain &ldquo;score&rdquo; and &ldquo;mapq&rdquo; columns.</p>
</td></tr>
<tr valign="top"><td><code>...</code></td>
<td>
<p>The rest of the arguments are passed to the <code>read.table()</code> function.</p>
</td></tr>
</table>


<h3>Value</h3>

<table summary="R valueblock">
<tr valign="top"><td><code>lib</code></td>
<td>
<p>Library name (factor).</p>
</td></tr>
<tr valign="top"><td><code>V</code></td>
<td>
<p>V segment name (factor).</p>
</td></tr>
<tr valign="top"><td><code>J</code></td>
<td>
<p>J segment name (factor).</p>
</td></tr>
<tr valign="top"><td><code>score</code></td>
<td>
<p>Alignment score (numeric).</p>
</td></tr>
<tr valign="top"><td><code>mapq</code></td>
<td>
<p>Mapping quality (numeric).  A sequence with a good alignment score will still have a low mapping quality if there are good alternative alignments to other V segments.</p>
</td></tr>
<tr valign="top"><td><code>read</code></td>
<td>
<p>Sequence read identifier (character).</p>
</td></tr>
<tr valign="top"><td><code>dna</code></td>
<td>
<p>DNA sequence of the CDR3 region (character).</p>
</td></tr>
<tr valign="top"><td><code>qual</code></td>
<td>
<p>Quality values for the DNA sequence (character).</p>
</td></tr>
<tr valign="top"><td><code>pep</code></td>
<td>
<p>Translation of the DNA sequence (character).</p>
</td></tr>
<tr valign="top"><td><code>unproductive</code></td>
<td>
<p>Flag indicating stop codons or frame shifts (logical).</p>
</td></tr>
<tr valign="top"><td><code>ambiguous</code></td>
<td>
<p>Flag indicating that the DNA sequences has ambiguous (&ldquo;N&rdquo;) nucleotides (logical).</p>
</td></tr>
</table>


<h3>Author(s)</h3>

<p>Charles Plessy</p>


<h3>See Also</h3>

<p><code><a href="clonotype_table.html">clonotype_table</a></code>, <code><a href="is_unproductive.html">is_unproductive</a></code>,
<code><a href="../../utils/html/read.table.html">read.table</a></code>,
Order Switchable Column Table (OSCT, <a href="http://sourceforge.net/projects/osctf/">http://sourceforge.net/projects/osctf/</a>)
</p>


<h3>Examples</h3>

<pre>
clonotypes &lt;- read_clonotypes(system.file('extdata', 'clonotypes.txt.gz', package = "clonotypeR"))
</pre>

<hr><div align="center">[Package <em>clonotypeR</em> version 1.1.3 <a href="00Index.html">Index</a>]</div>
</body></html>
